File: backend/models/Plot.js
const mongoose = require('mongoose');

const plotSchema = new mongoose.Schema({
  title: {
    type: String,
    required: [true, 'Plot title is required'],
    trim: true,
    maxlength: [100, 'Title cannot exceed 100 characters']
  },
  plotType: {
    type: String,
    required: true,
    enum: ['scatter', 'line', 'bar', 'histogram'],
    lowercase: true
  },
  datasetId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Dataset',
    required: true
  },
  xAxis: {
    type: String,
    required: [true, 'X-axis column is required']
  },
  yAxis: {
    type: String,
    required: function() {
      return ['scatter', 'line'].includes(this.plotType);
    }
  },
  filters: {
    type: mongoose.Schema.Types.Mixed,
    default: {}
  },
  createdBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  }
}, {
  timestamps: true
});

// Index for better query performance
plotSchema.index({ createdBy: 1, createdAt: -1 });
plotSchema.index({ datasetId: 1 });

module.exports = mongoose.model('Plot', plotSchema);

File: backend/app/models/plot.py
from pydantic import BaseModel, Field, ConfigDict
from typing import Optional, Dict, Any
from datetime import datetime
from bson import ObjectId
from app.models.user import PyObjectId

class PlotBase(BaseModel):
    title: str
    plot_type: str  # scatter, line, bar, histogram, etc.
    x_axis: str
    y_axis: Optional[str] = None
    config: Dict[str, Any] = Field(default_factory=dict)

class PlotCreate(PlotBase):
    dataset_id: PyObjectId

class PlotUpdate(BaseModel):
    title: Optional[str] = None
    plot_type: Optional[str] = None
    x_axis: Optional[str] = None
    y_axis: Optional[str] = None
    config: Optional[Dict[str, Any]] = None

class Plot(PlotBase):
    model_config = ConfigDict(arbitrary_types_allowed=True, populate_by_name=True)
    
    id: Optional[PyObjectId] = Field(default=None, alias="_id")
    dataset_id: PyObjectId
    user_id: PyObjectId
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
